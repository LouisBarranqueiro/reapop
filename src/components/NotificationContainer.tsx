import React, {ReactNode, useEffect, useState} from 'react'
import {Timer} from '../utils'
import {Notification as NotificationType} from '../reducers/notifications/types'
import {DismissNotification} from './NotificationsSystem'

type Props = {
    notification: NotificationType
    dismissNotification: DismissNotification
    children?: ReactNode
}

const NotificationContainer = (props: Props) => {
    const {notification, dismissNotification} = props
    const {dismissAfter, onAdd, onDismiss} = notification
    const [timer, setTimer] = useState<Timer | null>(null)

    useEffect(() => {
        if (onAdd) {
            onAdd()
        }
        return () => {
            if (onDismiss) {
                onDismiss()
            }
        }
    }, [])

    useEffect(() => {
        if (!timer && dismissAfter && dismissAfter > 0) {
            const timer = new Timer(dismissAfter, () => dismissNotification(notification.id))
            timer.resume()
            setTimer(timer)
        } else if (timer && !dismissAfter) {
            timer.pause()
            setTimer(null)
        }
    }, [dismissAfter])

    return (
        <div
            data-testid="timed-notification"
            onMouseEnter={timer ? () => timer.pause() : undefined}
            onMouseLeave={timer ? () => timer.resume() : undefined}
        >
            {props.children}
        </div>
    )
}

export default NotificationContainer
