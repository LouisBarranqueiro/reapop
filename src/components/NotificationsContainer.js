import React, {Component} from 'react';
import TransitionGroup from 'react/lib/ReactCSSTransitionGroup';
import Notification from './Notification';
import {POSITIONS} from '../constants';

export class NotificationsContainer extends Component {
  // default types
  static defaultProps = {
    notifications: []
  };

  // Properties types
  static propTypes = {
    notifications: React.PropTypes.array.isRequired,
    position: React.PropTypes.string.isRequired,
    theme: React.PropTypes.shape({
      notificationsContainer: React.PropTypes.shape({
        className: React.PropTypes.shape({
          main: React.PropTypes.string.isRequired,
          position: React.PropTypes.func.isRequired
        }).isRequired,
        transition: React.PropTypes.shape({
          name: React.PropTypes.object.isRequired,
          enterTimeout: React.PropTypes.number.isRequired,
          leaveTimeout: React.PropTypes.number.isRequired
        }).isRequired
      }).isRequired,
      notification: React.PropTypes.shape({
        className: React.PropTypes.object.isRequired
      }).isRequired
    }).isRequired
  };

  /**
   * Constructor
   * Bind methods
   * @param {Object} props
   * @returns {void}
   */
  constructor(props) {
    super(props);
    this._renderNotifications = this._renderNotifications.bind(this);
  }

  /**
   * Render notifications
   * @private
   * @returns {XML}
   */
  _renderNotifications() {
    // get all notifications and default values for notifications
    const {
      position,
      theme: {
        notification: {className}
      }
    } = this.props;
    let {notifications} = this.props;

    // when notifications are displayed at the bottom,
    // we display notifications from bottom to top
    if ([POSITIONS.bottomLeft, POSITIONS.bottomRight, POSITIONS.bottomCenter, POSITIONS.bottom]
        .indexOf(position) >= 0) {
      notifications = notifications.reverse();
    }

    return notifications.map((notification) => (
      <Notification key={notification.id} notification={notification} className={className}/>
    ));
  }

  /**
   * Render
   * @returns {XML}
   */
  render() {
    const {
      className, transition: {name, enterTimeout, leaveTimeout}
    } = this.props.theme.notificationsContainer;
    const {position} = this.props;

    return (
      <div className={`${className.main} ${className.position(position)}`}>
        <TransitionGroup
          transitionName={name}
          transitionEnterTimeout={enterTimeout}
          transitionLeaveTimeout={leaveTimeout}>
          {this._renderNotifications()}
        </TransitionGroup>
      </div>
    );
  }
}

export default NotificationsContainer;
