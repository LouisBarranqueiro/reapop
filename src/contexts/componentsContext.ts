import {createContext} from 'react'
import NotificationIcon from '../components/NotificationIcon'
import Notification from '../components/Notification'
import SlideTransition from '../components/SlideTransition'
import GrowTransition from '../components/GrowTransition'
import FadeTransition from '../components/FadeTransition'

export type ComponentContextType = {
    Transition?: typeof SlideTransition | typeof GrowTransition | typeof FadeTransition
    NotificationIcon?: typeof NotificationIcon
    Notification?: typeof Notification
}

export const ComponentsContext = createContext<ComponentContextType>({})
