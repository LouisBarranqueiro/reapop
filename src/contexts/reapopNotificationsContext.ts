import {createContext} from 'react'
import {NewNotification, Notification} from '../reducers/notifications/types'

type NotificationsContext = {
    notifications: Notification[]
    notify: (notification: NewNotification) => void
    dismissNotification: (id: string) => void
    dismissNotifications: () => void
}

export const ReapopNotificationsContext = createContext<NotificationsContext>(undefined!)
