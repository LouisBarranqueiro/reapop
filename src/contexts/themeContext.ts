import {createContext} from 'react'
import {Theme} from '..'

export const ThemeContext = createContext<Theme | undefined>(undefined)
