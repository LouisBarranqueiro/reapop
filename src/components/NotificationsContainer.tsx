import React from 'react'
import {TransitionGroup} from 'react-transition-group'
import {Notification as NotificationType, Position} from '../reducers/notifications/types'
import {classnames, POSITIONS} from '../constants'
import {useComponentsContext} from '../hooks/useComponentsContext'
import {useTheme} from '../hooks/useTheme'
import SlideTransition from './SlideTransition'
import NotificationContainer from './NotificationContainer'
import NotificationComponent from './Notification'
import {DismissNotification} from './NotificationsSystem'

type Props = {
    notifications: NotificationType[]
    position: Position
    dismissNotification: DismissNotification
    singleContainer: boolean
}

const NotificationsContainer = (props: Props) => {
    const {position, dismissNotification, singleContainer} = props
    const components = useComponentsContext()
    const theme = useTheme()
    const styles = theme ? theme.container(position, singleContainer) : {}
    const classname = classnames.container(position, singleContainer)
    const Transition = components.Transition || SlideTransition
    const Notification = components.Notification || NotificationComponent
    let {notifications} = props

    // when notifications are displayed at the bottom,
    // we display notifications from bottom to top
    if (([POSITIONS.bottomRight, POSITIONS.bottomCenter, POSITIONS.bottomLeft] as string[]).includes(position)) {
        notifications = notifications.slice().reverse()
    }

    return (
        <div style={styles} className={classname.join(' ')}>
            <TransitionGroup>
                {notifications.map((notification) => (
                    <Transition key={notification.id} notification={notification}>
                        <NotificationContainer
                            key={notification.id}
                            notification={notification}
                            dismissNotification={dismissNotification}
                        >
                            <Notification
                                notification={notification}
                                dismissNotification={dismissNotification}
                                theme={theme}
                                components={components}
                            />
                        </NotificationContainer>
                    </Transition>
                ))}
            </TransitionGroup>
        </div>
    )
}

export default NotificationsContainer
