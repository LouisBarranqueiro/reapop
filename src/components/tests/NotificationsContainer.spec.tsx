import React, {ReactNode} from 'react'
import pretty from 'pretty'
import {render} from '@testing-library/react'

import {Notification, POSITIONS, STATUSES, Theme} from '../..'
import {ComponentContextType, ComponentsContext} from '../../contexts/componentsContext'
import {ThemeContext} from '../../contexts/themeContext'
import atalhoTheme from '../../themes/atalho'

import NotificationsContainer from '../NotificationsContainer'
import NotificationContainer from '../NotificationContainer'

jest.mock('../NotificationContainer', () => jest.fn(({notification}) => <div>notification-{notification.id}</div>))

describe('<NotificationsContainer/>', () => {
    const notifications: Notification[] = [
        {
            id: '1',
            position: POSITIONS.topLeft,
            status: STATUSES.none,
            buttons: [],
        },
        {
            id: '2',
            position: POSITIONS.topLeft,
            status: STATUSES.none,
            buttons: [],
        },
    ]

    const renderWithContexts = (children: ReactNode, theme?: Theme, components: ComponentContextType = {}) =>
        render(
            <ComponentsContext.Provider value={components}>
                <ThemeContext.Provider value={theme}>{children}</ThemeContext.Provider>
            </ComponentsContext.Provider>
        )

    beforeEach(() => {
        jest.spyOn(atalhoTheme, 'container').mockReturnValue({color: 'red'})
    })

    afterEach(() => {
        jest.clearAllMocks()
    })

    afterAll(() => {
        jest.resetAllMocks()
    })

    it.each([
        ['without theme', undefined],
        ['with theme', atalhoTheme],
    ])('should render notifications %s', (_, theme) => {
        const dismissNotification = jest.fn()
        const {container} = renderWithContexts(
            <NotificationsContainer
                notifications={notifications}
                position={notifications[0].position}
                dismissNotification={dismissNotification}
                singleContainer={false}
            />,
            theme
        )

        expect(container.innerHTML).toMatchSnapshot()
        expect(NotificationContainer).toMatchSnapshot()
    })

    it.each(Object.values(POSITIONS))('should display notifications for %s position', (position) => {
        const {container} = renderWithContexts(
            <NotificationsContainer
                notifications={notifications}
                position={position}
                dismissNotification={jest.fn()}
                singleContainer
            />
        )
        expect(pretty(container.innerHTML)).toMatchSnapshot()
    })

    it('should display notifications with custom components', () => {
        const CustomTransition = jest.fn((props) => <div id="custom-transition">{props.children}</div>)
        const CustomNotification = jest.fn(() => <div id="custom-notification"></div>)

        const {container} = renderWithContexts(
            <NotificationsContainer
                notifications={notifications}
                position={notifications[0].position}
                dismissNotification={jest.fn()}
                singleContainer
            />,
            undefined,
            {Transition: CustomTransition, Notification: CustomNotification}
        )
        expect(pretty(container.innerHTML)).toMatchSnapshot()
        expect(CustomTransition.mock.calls).toMatchSnapshot()
        expect(CustomNotification.mock.calls).toMatchSnapshot()
    })
})
