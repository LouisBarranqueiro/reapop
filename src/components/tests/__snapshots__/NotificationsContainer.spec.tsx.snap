// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotificationsContainer/> should display notifications for bottom-center position 1`] = `
"<div class=\\"reapop__container reapop__container--bottom-center reapop__container--single\\">
  <div>
    <div>notification-2</div>
    <div>notification-1</div>
  </div>
</div>"
`;

exports[`<NotificationsContainer/> should display notifications for bottom-left position 1`] = `
"<div class=\\"reapop__container reapop__container--bottom-left reapop__container--single\\">
  <div>
    <div>notification-2</div>
    <div>notification-1</div>
  </div>
</div>"
`;

exports[`<NotificationsContainer/> should display notifications for bottom-right position 1`] = `
"<div class=\\"reapop__container reapop__container--bottom-right reapop__container--single\\">
  <div>
    <div>notification-2</div>
    <div>notification-1</div>
  </div>
</div>"
`;

exports[`<NotificationsContainer/> should display notifications for top-center position 1`] = `
"<div class=\\"reapop__container reapop__container--top-center reapop__container--single\\">
  <div>
    <div>notification-1</div>
    <div>notification-2</div>
  </div>
</div>"
`;

exports[`<NotificationsContainer/> should display notifications for top-left position 1`] = `
"<div class=\\"reapop__container reapop__container--top-left reapop__container--single\\">
  <div>
    <div>notification-1</div>
    <div>notification-2</div>
  </div>
</div>"
`;

exports[`<NotificationsContainer/> should display notifications for top-right position 1`] = `
"<div class=\\"reapop__container reapop__container--top-right reapop__container--single\\">
  <div>
    <div>notification-1</div>
    <div>notification-2</div>
  </div>
</div>"
`;

exports[`<NotificationsContainer/> should display notifications with custom components 1`] = `
"<div class=\\"reapop__container reapop__container--top-left reapop__container--single\\">
  <div>
    <div id=\\"custom-transition\\">
      <div>notification-1</div>
    </div>
    <div id=\\"custom-transition\\">
      <div>notification-2</div>
    </div>
  </div>
</div>"
`;

exports[`<NotificationsContainer/> should display notifications with custom components 2`] = `
Array [
  Array [
    Object {
      "appear": undefined,
      "children": <mockConstructor
        dismissNotification={[MockFunction]}
        notification={
          Object {
            "buttons": Array [],
            "id": "1",
            "position": "top-left",
            "status": "none",
          }
        }
      >
        <mockConstructor
          components={
            Object {
              "Notification": [MockFunction],
              "Transition": [MockFunction] {
                "calls": [Circular],
                "results": Array [
                  Object {
                    "type": "return",
                    "value": <div
                      id="custom-transition"
                    >
                      <mockConstructor
                        dismissNotification={[MockFunction]}
                        notification={
                          Object {
                            "buttons": Array [],
                            "id": "1",
                            "position": "top-left",
                            "status": "none",
                          }
                        }
                      >
                        <mockConstructor
                          components={[Circular]}
                          dismissNotification={[MockFunction]}
                          notification={
                            Object {
                              "buttons": Array [],
                              "id": "1",
                              "position": "top-left",
                              "status": "none",
                            }
                          }
                        />
                      </mockConstructor>
                    </div>,
                  },
                  Object {
                    "type": "return",
                    "value": <div
                      id="custom-transition"
                    >
                      <mockConstructor
                        dismissNotification={[MockFunction]}
                        notification={
                          Object {
                            "buttons": Array [],
                            "id": "2",
                            "position": "top-left",
                            "status": "none",
                          }
                        }
                      >
                        <mockConstructor
                          components={[Circular]}
                          dismissNotification={[MockFunction]}
                          notification={
                            Object {
                              "buttons": Array [],
                              "id": "2",
                              "position": "top-left",
                              "status": "none",
                            }
                          }
                        />
                      </mockConstructor>
                    </div>,
                  },
                ],
              },
            }
          }
          dismissNotification={[MockFunction]}
          notification={
            Object {
              "buttons": Array [],
              "id": "1",
              "position": "top-left",
              "status": "none",
            }
          }
        />
      </mockConstructor>,
      "enter": undefined,
      "exit": undefined,
      "in": true,
      "notification": Object {
        "buttons": Array [],
        "id": "1",
        "position": "top-left",
        "status": "none",
      },
      "onExited": [Function],
    },
    Object {},
  ],
  Array [
    Object {
      "appear": undefined,
      "children": <mockConstructor
        dismissNotification={[MockFunction]}
        notification={
          Object {
            "buttons": Array [],
            "id": "2",
            "position": "top-left",
            "status": "none",
          }
        }
      >
        <mockConstructor
          components={
            Object {
              "Notification": [MockFunction],
              "Transition": [MockFunction] {
                "calls": [Circular],
                "results": Array [
                  Object {
                    "type": "return",
                    "value": <div
                      id="custom-transition"
                    >
                      <mockConstructor
                        dismissNotification={[MockFunction]}
                        notification={
                          Object {
                            "buttons": Array [],
                            "id": "1",
                            "position": "top-left",
                            "status": "none",
                          }
                        }
                      >
                        <mockConstructor
                          components={[Circular]}
                          dismissNotification={[MockFunction]}
                          notification={
                            Object {
                              "buttons": Array [],
                              "id": "1",
                              "position": "top-left",
                              "status": "none",
                            }
                          }
                        />
                      </mockConstructor>
                    </div>,
                  },
                  Object {
                    "type": "return",
                    "value": <div
                      id="custom-transition"
                    >
                      <mockConstructor
                        dismissNotification={[MockFunction]}
                        notification={
                          Object {
                            "buttons": Array [],
                            "id": "2",
                            "position": "top-left",
                            "status": "none",
                          }
                        }
                      >
                        <mockConstructor
                          components={[Circular]}
                          dismissNotification={[MockFunction]}
                          notification={
                            Object {
                              "buttons": Array [],
                              "id": "2",
                              "position": "top-left",
                              "status": "none",
                            }
                          }
                        />
                      </mockConstructor>
                    </div>,
                  },
                ],
              },
            }
          }
          dismissNotification={[MockFunction]}
          notification={
            Object {
              "buttons": Array [],
              "id": "2",
              "position": "top-left",
              "status": "none",
            }
          }
        />
      </mockConstructor>,
      "enter": undefined,
      "exit": undefined,
      "in": true,
      "notification": Object {
        "buttons": Array [],
        "id": "2",
        "position": "top-left",
        "status": "none",
      },
      "onExited": [Function],
    },
    Object {},
  ],
]
`;

exports[`<NotificationsContainer/> should display notifications with custom components 3`] = `Array []`;

exports[`<NotificationsContainer/> should render notifications with theme 1`] = `"<div style=\\"color: red;\\" class=\\"reapop__container reapop__container--top-left\\"><div><div>notification-1</div><div>notification-2</div></div></div>"`;

exports[`<NotificationsContainer/> should render notifications with theme 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "children": <Notification
          components={Object {}}
          dismissNotification={[MockFunction]}
          notification={
            Object {
              "buttons": Array [],
              "id": "1",
              "position": "top-left",
              "status": "none",
            }
          }
          theme={
            Object {
              "container": [MockFunction] {
                "calls": Array [
                  Array [
                    "top-left",
                    false,
                  ],
                ],
                "results": Array [
                  Object {
                    "type": "return",
                    "value": Object {
                      "color": "red",
                    },
                  },
                ],
              },
              "notification": [Function],
              "notificationButton": [Function],
              "notificationButtonText": [Function],
              "notificationButtons": [Function],
              "notificationDismissIcon": [Function],
              "notificationIcon": [Function],
              "notificationImage": [Function],
              "notificationImageContainer": [Function],
              "notificationMessage": [Function],
              "notificationMeta": [Function],
              "notificationTitle": [Function],
            }
          }
        />,
        "dismissNotification": [MockFunction],
        "notification": Object {
          "buttons": Array [],
          "id": "1",
          "position": "top-left",
          "status": "none",
        },
      },
      Object {},
    ],
    Array [
      Object {
        "children": <Notification
          components={Object {}}
          dismissNotification={[MockFunction]}
          notification={
            Object {
              "buttons": Array [],
              "id": "2",
              "position": "top-left",
              "status": "none",
            }
          }
          theme={
            Object {
              "container": [MockFunction] {
                "calls": Array [
                  Array [
                    "top-left",
                    false,
                  ],
                ],
                "results": Array [
                  Object {
                    "type": "return",
                    "value": Object {
                      "color": "red",
                    },
                  },
                ],
              },
              "notification": [Function],
              "notificationButton": [Function],
              "notificationButtonText": [Function],
              "notificationButtons": [Function],
              "notificationDismissIcon": [Function],
              "notificationIcon": [Function],
              "notificationImage": [Function],
              "notificationImageContainer": [Function],
              "notificationMessage": [Function],
              "notificationMeta": [Function],
              "notificationTitle": [Function],
            }
          }
        />,
        "dismissNotification": [MockFunction],
        "notification": Object {
          "buttons": Array [],
          "id": "2",
          "position": "top-left",
          "status": "none",
        },
      },
      Object {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": <div>
        notification-
        1
      </div>,
    },
    Object {
      "type": "return",
      "value": <div>
        notification-
        2
      </div>,
    },
  ],
}
`;

exports[`<NotificationsContainer/> should render notifications without theme 1`] = `"<div class=\\"reapop__container reapop__container--top-left\\"><div><div>notification-1</div><div>notification-2</div></div></div>"`;

exports[`<NotificationsContainer/> should render notifications without theme 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "children": <Notification
          components={Object {}}
          dismissNotification={[MockFunction]}
          notification={
            Object {
              "buttons": Array [],
              "id": "1",
              "position": "top-left",
              "status": "none",
            }
          }
        />,
        "dismissNotification": [MockFunction],
        "notification": Object {
          "buttons": Array [],
          "id": "1",
          "position": "top-left",
          "status": "none",
        },
      },
      Object {},
    ],
    Array [
      Object {
        "children": <Notification
          components={Object {}}
          dismissNotification={[MockFunction]}
          notification={
            Object {
              "buttons": Array [],
              "id": "2",
              "position": "top-left",
              "status": "none",
            }
          }
        />,
        "dismissNotification": [MockFunction],
        "notification": Object {
          "buttons": Array [],
          "id": "2",
          "position": "top-left",
          "status": "none",
        },
      },
      Object {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": <div>
        notification-
        1
      </div>,
    },
    Object {
      "type": "return",
      "value": <div>
        notification-
        2
      </div>,
    },
  ],
}
`;
